zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/CallChannel.kt

```diff
@@ -101,3 +97,36 @@ private object ThrowableSerializer : KSerializer<Throwable> {
 private class ThrowableSurrogate(
   val message: String
 )
+
+/**
+ * This is a special serializer because it's scoped to an endpoint. It is not a general-purpose
+ * serializer and only works in Zipline.
+ *
+ * To send a handle to an inbound service, we register it with the endpoint and transmit the
+ * registered identifier.
+ *
+ * To receive a handle, we record the received identifier and return it when making calls against
+ * the referenced service.
+ */
+internal class HandleSerializer<T : Any>(
+  val endpoint: Endpoint
+) : KSerializer<Handle<T>> {
+  override val descriptor = PrimitiveSerialDescriptor("Handle", PrimitiveKind.STRING)
+
+  override fun serialize(encoder: Encoder, value: Handle<T>) {
+    val name = endpoint.generateName()
+    if (value is InboundHandle<T>) {
+      value.connect(endpoint, name)
+      encoder.encodeString(name)
+    } else {
+      error("serializing an outbound handle is not implemented")
```

Comment:
There’s a fancy implementation possible here. . . if you send a Flow back to its origin we can send it back by reference and not by value.